streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+            // expected end state per output partition:
+            // p-0: 10 records + COMMIT + 5 records + COMMIT + 5 records + COMMIT + 5 records (+ maybe commit if tasks would have been switch; we don't care and don't verify)
+            // p-1: 10 records + COMMIT + 5 records + 5 records + COMMIT  + 5 records + COMMIT
+            commitRequested.set(0);
+            stateTransitions1.clear();
+            stateTransitions2.clear();
+            streams2Beta.start();
+            waitForCondition(
+                () -> stateTransitions1.equals(Arrays.asList(
+                    KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.REBALANCING),
+                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING),
+                    KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.REBALANCING),
+                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING)
+                )) && stateTransitions2.equals(Arrays.asList(
+                    KeyValue.pair(KafkaStreams.State.CREATED, KafkaStreams.State.REBALANCING),
+                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING),
+                    KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.REBALANCING),
+                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING)
+                )),
+                MAX_WAIT_TIME_MS,
+                "Clients did not startup and stabilize on time."
+            );
+
+            // capture committed state
+            final Map<Long, Long> currentStateAfterSecondUpgrade = new HashMap<>(currentStateOne);
+            try (final KeyValueIterator<Long, Long> it = storeOneAfterUpgrade.all()) {
+                while (it.hasNext()) {
+                    final KeyValue<Long, Long> row = it.next();
+                    if (row.key == keyFilterSecondUpgrade) {
+                        currentStateAfterSecondUpgrade.put(row.key, row.value);
+                    }
+                }
+            }
+            final ReadOnlyKeyValueStore<Long, Long> storeTwoAfterUpgrade =
+                streams2Beta.store(StoreQueryParameters.fromNameAndType(storeName, QueryableStoreTypes.keyValueStore()));
+            try (final KeyValueIterator<Long, Long> it = storeTwoAfterUpgrade.all()) {
+                while (it.hasNext()) {
+                    final KeyValue<Long, Long> row = it.next();
+                    if (row.key == keyFilterSecondUpgrade) {
+                        currentStateAfterSecondUpgrade.put(row.key, row.value);
+                    }
+                }
+            }
+
+            // phase 10:
+            // expected end state per output partition:
+            // p-0: 10 records + COMMIT + 5 records + COMMIT + 5 records + COMMIT + 5 records + 5 records + COMMIT
+            // p-1: 10 records + COMMIT + 5 records + 5 records + COMMIT  + 5 records + COMMIT + 5 records + COMMIT
```

Comment:
This is a meta comment: I think the key here is to check that the offsets sent via "sendOffsetsToTxn" would not be seen to the new owner of the task, i.e. the fencing would take effects. But here since we do not commit (only 5 out of 10 records) this logic would never be checked.

On the other hand, inside StreamsProducer we always call 

```
            producer.sendOffsetsToTransaction(offsets, consumerGroupMetadata);
            producer.commitTransaction();
```

together, it is a bit hard to check with the given producer.

I'd suggest doing a bit different here: we use a customized client-supplier that generate an extended producer which, depending on the failure injection boolean, throws an exception in `commitTransaction` immediately, so that we made sure the `sendOffsetsToTransaction` completes (i.e. the request acked from group coordinator already) before we crash, instead of crashing in the middle of the processing where we'd not send any offsets to group coordinator anyways.

Reply:
Correct. I was aware that this case would not be covered, but did not have a good idea _how_ to test it. Your idea is great! Will work on that.